### PR TITLE
#30 add api interactions in post route

### PR DESCRIPTION
**Pull request by Harrisman05**
### Feature Request link
[#30](https://github.com/orgs/alphagov/projects/65/views/1?pane=issue&itemId=29550285)

### Change description

- Add fetch request to [postcodes.io](https://postcodes.io/) to retrieve parliamentary constituency in a given postcode
- Add fetch request to [members.api](https://members-api.parliament.uk/index.html) to retrieve elected representative using parliamentary constituency.
- Add TypeScript interfaces for expected response from postcodes.io and members-api
- Add very basic in-line styling to display the elected representative clearly.

### CODE REVIEW

**Review by huwd — APPROVED** [review]
This looks good - if you're up for it I have three major thoughts:

1. Code is good but we can look at the commits here, there's a bit of reshuffling between them. Try going to the commit page here and stepping through there. "Do your commits tell a story?" This probably isn't worth fixing on this PR but... I'd say you've really got one change here so you might be ok to do `git reabse -i main` and squash / fixup down into one commit and then broaden the commit label.

2. There's also a case for this not quite being a full Pull Request yet. I note that there's still some work to do here, and great that we were able to get this together on Friday in time for a review. But don't necessarily need to merge this now if you'd intended to pick this up next Friday and continue on this branch. This might be down to me asking for PRs to review end of Fridays! So perhaps... the usual rule of thumb should apply. Only PRs when we're fully ready to merge into main and we want to be looking at completed or nice isolated chunks of features to do that. BUT if you want an early review we can open a draft PR and you can get early feedback based on however far you got on a Friday. and then you could pick up and keep working on that branch until you hit that nice logical unit of code to merge in?

3. Again code is great but there are likely some savings you can make here. A couple of console.logs as noted. Big one are those huge interfaces. Those look like a really comprehensive overview of everything you can get off the two APIs. But, remember you don't need to annotate everything you get back here, you only need to provide types for the things your application would actually use.

So for instance given an API response:
```
response = {
  "a": 1,
  "b": 2,
  "c": { "foo": "bar }
}
```

if the only thing my app was ever going to call was `response["b"].

Then my interface only ever needs to be:

```
interface Response {
  "b": number
}
```

yes there is much more in the JSON response, but if you're not using it, typescript doesn't need to know about it. This mostly saves you the upfront work that you've already done, but could also save you a few lines of code.

I understand the impulse though, i found myself doing this a lot with typescript until I discovered things like OpenAPI. For me it scratched that itch to get [the documentation](https://editor.swagger.io/) I wanted to give me the overview of "what can I use", and then let me keep the types as just that safety check for things my app actually consumes (essentially which bits of the general API response does _this_ app actually care about).

Especially as JSON responses are coming in from outside the app, and so are always really an `any` time. You are then telling typescript "right from this point on, pretend that they can't just be anything and I'm promising everything you'll get back from now on out is a `PostcodeApiResponse` in this format"

Then it can check everything works logically before we compile it into javascript and lose all these interfaces anyway.

Hope that helps?

**Comment by Harrisman05** [comment]
Thanks for the in-depth comments!

1) Yea, I probably agree the commits can be squashed down, but I just wanted to commit them as atomic as possible to start with so you could have a look. I want to practice rebasing too... 

There are a few temp commits in here that I didn't bother to rebase at the time too. I am not very great at determining how long/short a commit should be still, so that's something I'm trying to improve. I would think there should be 2 commits tho: _add API interactions in POST route_ and _add basic styling to `index.njk`_, 2 commits as one is backend and one is frontend... **Or should I just squash the frontend commit too as it's so minor?**

**Another question was if I need to run `git merge main`, should I always squash it if it's minor or leave it as a commit if it's a complicated merge?**

2) I would also agree, altho I did want to raise the PR just so I could get a review out of you. The API logic needs to be moved into it's own function as the controller/route is quite bloated atm. **Would separate functions for separate API calls be good?The fetching is obviously quite similar but the parsing for both is different... what do you think?**

3) Fully typing the postcode API response is something I did with Antoni to teach him about interfaces, and then I just copied his code. For the members API response, I got AI to make the interfaces its a good usecase and quite simple (altho would need to be validated IRL ofc). But yea, I will trim down the interfaces to only have properties for things I actually need atm.

In fact without any validation they are ofc just `any`, so at that point I could validate at runtime eg with `zod`, but this only a nice to have and not important for my MVP.

File: src/index.ts [hunk 1]
```diff
@@ -15,13 +16,38 @@ nunjucks.configure('views', {
...
+			responseJson.result.parliamentary_constituency;
+
+		const electedRepresentativeRes = await fetch(
+			`https://members-api.parliament.uk/api/Location/Constituency/Search?searchText=${parliamentaryConstituency}&skip=0&take=20`
+		);
+		const electedRepresentativeJson: MembersApiResponse = await electedRepresentativeRes.json();
+		const electedRepresentativeName =
+			electedRepresentativeJson.items[0].value.currentRepresentation.member
+				.value.nameFullTitle;
+		console.log(electedRepresentativeName);
```
[the remark this answers is not in the document]
Comment (Harrisman05) [reply src/index.ts:41]: Console.logs were just there for debugging in express server. I will remove them when I rebase.

File: src/index.ts [hunk 2]
```diff
@@ -15,13 +16,38 @@ nunjucks.configure('views', {
 
 app.get('/', (req: Request, res: Response) => {
 	console.log('Hello, client has sent get request');
-	res.render('index.njk');
+	res.render('index.njk', { electedRepresentativeName: '' });
```
[the remark this answers is not in the document]
Comment (Harrisman05) [reply src/index.ts:19]: Probably will add a placeholder value here instead of empty string, just hadn't decided what to render (could use html placeholder attribute instead in the frontend?)

**Comment by Harrisman05** [comment]
Original commit stack was rebased as it was hard to follow and outlined more of the "dev workflow" in reaching the end goal of implementing the APIs. 

Commit stack rebased in such a way that it documents the "official history" of the repo, is easier to read and understand for other developers.
